components/console/esp_linenoise.h

```diff
@@ -0,0 +1,237 @@
+/*
+ * SPDX-FileCopyrightText: 2016-2021 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "esp_err.h"
+#include <stdbool.h>
+#include <stddef.h>
+
+typedef struct {
+  size_t len;
+  char **cvec;
+} esp_linenoise_completions_t;
+
+// Forward define
+struct esp_linenoise_t;
+
+// Linenoise handle is a pointer to an internal structure
+typedef struct esp_linenoise_t* esp_linenoise_handle_t;
+
+typedef void(esp_linenoise_completion_callback_t)(const char *buf, esp_linenoise_completions_t *lc);
+typedef char*(esp_linenoise_hints_callback_t)(const char *buf, int *color, int *bold);
+typedef void(esp_linenoise_free_hints_callback_t)(void *buf);
+
+/**
+ * @brief Create a new linenoise instance.
+ * 
+ * @returns A new linenoise handle.
+ */
+esp_linenoise_handle_t esp_linenoise_create();
+
+/**
+ * @brief The main API of the linenoise library.
+ * 
+ * @param prompt the command line prompt to show the user when ready to accept input.
+ * 
+ * @returns the console command the user entered, or NULL if failure.
+ */
+char* esp_linenoise(esp_linenoise_handle_t handle, const char *prompt);
```

Comment:
All other functions return esp_err_t, perhaps we could apply the same treatment to this one?

```suggestion
esp_err_t esp_linenoise_get_line(esp_linenoise_handle_t handle, const char *prompt, char** out_line);
```

Reply:
Yes I considered that strongly as well. I would be in favor of that.

Reply:
Preferably have the user pass the buffer. But we would need to handle the linenoise wrapper differently.

`esp_err_t esp_linenoise_get_line(esp_linenoise_handle_t handle, const char *prompt, char* out_line, size_t len);`